source/trace_event.c

```diff
@@ -0,0 +1,191 @@
+        }                                                                                                              \
+    } while (0)
+
+struct aws_trace_event {
+    struct cJSON *root, *event_array;
+    struct aws_bus bus;
+    struct aws_allocator *allocator;
+};
+
+/* Add trace event meta data to JSON object when the message bus allows it */
+void aws_trace_event_listener(uint64_t address, const void *msg, void *user_data) {
+    struct aws_trace_event_metadata *trace_event_data = (struct aws_trace_event_metadata *)msg;
+    if (trace_event_data == NULL) {
+        aws_raise_error(AWS_ERROR_OOM);
+    }
+    cJSON *event = cJSON_CreateObject();
+    if (event == NULL) {
+        aws_raise_error(AWS_ERROR_OOM);
+    }
+
+    cJSON_AddItemToArray(trace_event->event_array, event);
+
+    /* add options for args later on, will initially be empty for now */
+    cJSON_AddItemToObject(event, "args", cJSON_CreateObject());
+
+    /* add more options later on */
+    if (cJSON_AddStringToObject(event, "cat", trace_event_data->category) == NULL) {
+        aws_raise_error(AWS_ERROR_OOM);
+    }
+    if (cJSON_AddStringToObject(event, "name", trace_event_data->name) == NULL) {
+        aws_raise_error(AWS_ERROR_OOM);
+    }
+    /* Fix for buffer overflow issue using char phase reference */
+    char ph[2];
+    strncat(ph, &(trace_event_data->phase), 1);
+    if (cJSON_AddStringToObject(event, "ph", ph) == NULL) {
+        aws_raise_error(AWS_ERROR_OOM);
+    }
+    if (cJSON_AddNumberToObject(event, "pid", trace_event_data->process_id) == NULL) {
+        aws_raise_error(AWS_ERROR_OOM);
+    }
+    if (cJSON_AddNumberToObject(event, "tid", trace_event_data->thread_id)) {
+        aws_raise_error(AWS_ERROR_OOM);
+    }
+    if (cJSON_AddNumberToObject(event, "ts", (double)trace_event_data->timestamp)) {
+        aws_raise_error(AWS_ERROR_OOM);
+    }
+}
+
+/* Destructor of trace event metadata sent through the bus */
+void aws_trace_event_destroy(void *payload) {}
+
+void aws_trace_system_write(int time_unit, const char *filename) {
+    if (time_unit) {
+        if (cJSON_AddStringToObject(trace_event->root, "displayTimeUnit", "ns") == NULL) {
+            aws_raise_error(AWS_ERROR_OOM);
+        }
+    }
+    char *out = cJSON_Print((trace_event->root));
+    if (out == NULL) {
+        aws_raise_error(AWS_ERROR_OOM);
+    }
+    FILE *fp;
+    char fn[strlen(filename) + 6];
+
+    strcpy(fn, filename);
+    strncat(fn, ".json", 6);
+
+    fp = fopen(fn, "w");
+    if (fp == NULL) {
+        aws_raise_error(AWS_ERROR_OOM);
+    }
+    fprintf(fp, "%s", out);
+    fclose(fp);
+    aws_mem_release(trace_event->allocator, out);
+}
+
+/*
+ * Public API
+ */
+
+/* starts the message bus and initializes the JSON root, and the event array for storing metadata */
+int aws_trace_system_init(struct aws_allocator *allocator) {
+    if (allocator == NULL) {
+        return AWS_OP_ERR;
```

Comment:
Need to be sure the reason for the failure is set before -1 is returned
`return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);`